gist.go

```diff
@@ -0,0 +1,86 @@
+package model
+
+// Typed constants
+type ChannelType string
+
+// CamelCase Constant Names
+const (
+	ChannelTypeOpen    ChannelType = "O"
+	ChannelTypePrivate ChannelType = "P"
+)
+
+type Channel struct {
+	// Member variables are not exported to ensure immutability.
+	id       string `json:"id"`
+	name     string `json:"name"`
+	createAt int64  `json:"create_at"`
+	// etc...
+
+	// Question: what do we do when there's a semantically meaningful
+	// "null" value for a member variable?
+	schemeID *string `json:"scheme_id"`
+
+	// Question: is this the best way of storing props and similar map/list data?
+	props map[string]interface{} `json:"props" db:"-"`
+}
+
+// Getters to access the member variables are provided. These *never* return pointers.
+// Correct capitalisation of initialisms
+func (channel *Channel) GetID() string {
+	return channel.id
+}
+
+// Returning `interface{}` feels gross. Can we do better than this?
+func (channel *Channel) GetProp(key string) interface{} {
+	return channel.props[key]
+}
+
+// "Patch" object holds changes to the object to make.
+type ChannelPatch struct {
+	Name *string `json:"name"`
+	// etc...
+
+	// Question: can we avoid pointers here and have some type like "NullableString"
+	// that indicates whether the field should be set explicitly instead?
+}
+
+// Similarly to what we do for the "PATCH" API endpoints currently, we modify model
+// objects by populating a "patch" object then calling a "ApplyPatch" method.
+func (channel *Channel) ApplyPatch(patch ChannelPatch) Channel {
+	// This method forces a deep-copy of the object to ensure immutability.
+	newChannel := channel.DeepCopy()
+	if patch.Name != nil {
+		newChannel.name = *patch.Name
+	}
+
+	// Question: How do we handle setting / deletion of Props?
+
+	return newChannel
+}
+
+// DeepCopy methods for model objects ensure immutability.
+func (channel *Channel) DeepCopy() Channel {
+	// Question: can we automate the implementation of these DeepCopy functions?
```

Comment:
Sure we can with a generator. Probably we can auto-generate a big percentage of this functions (Getters, Patch struct, DeepCopy/ApplyPatch functions...

Probably we can build something interesting around this idea of generate. so we can define the struct itself, and all (or 90% of) the inmutability machinery is generated.

Reply:
We could also use a library like https://pkg.go.dev/github.com/getlantern/deepcopy

Reply:
:+1:. I like code generation!

Reply:
We also have this: https://github.com/globusdigital/deep-copy.